helix-lsp/src/snippet.rs

```diff
@@ -0,0 +1,507 @@
+                insert.push_str(&text);
+            }
+            Variable {
+                name: _name,
+                regex: None,
+                r#default,
+            } => {
+                // TODO: variables. For now, fall back to the default, which defaults to "".
+                let text = r#default.unwrap_or_default();
+                offset += text.chars().count();
+                insert.push_str(text);
+            }
+            Tabstop { tabstop } => {
+                tabstops.push((tabstop, Range::point(offset)));
+            }
+            Placeholder { tabstop, value } => match value.as_ref() {
+                // https://doc.rust-lang.org/beta/unstable-book/language-features/box-patterns.html
+                // would make this a bit nicer
+                Text(text) => {
+                    if include_placeholer {
+                        let len_chars = text.chars().count();
+                        tabstops.push((tabstop, Range::new(offset, offset + len_chars + 1)));
+                        offset += len_chars;
+                        insert.push_str(text);
+                    } else {
+                        tabstops.push((tabstop, Range::point(offset)));
+                    }
+                }
+                other => {
+                    log::error!(
+                        "Discarding snippet: generating a transaction for placeholder contents {:?} is unimplemented.",
+                        other
+                    );
+                    return Transaction::new(doc);
+                }
+            },
+            other => {
+                log::error!(
+                    "Discarding snippet: generating a transaction for {:?} is unimplemented.",
+                    other
+                );
+                return Transaction::new(doc);
+            }
+        }
+    }
+
+    let transaction = Transaction::change_by_selection(doc, selection, |range| {
+        let cursor = range.cursor(text);
+        (
+            (cursor as i128 + start_offset) as usize,
+            (cursor as i128 + end_offset) as usize,
+            Some(insert.clone().into()),
+        )
+    });
+
+    // sort in assending order (except for 0, which should always be the last one (per lsp doc))
+    tabstops.sort_unstable_by_key(|(n, _range)| if *n == 0 { usize::MAX } else { *n });
+
+    // merge tabstops with the same index (we take advantage of the fact that we just sorted them
+    // above to simply look backwards)
+    let mut ntabstops = Vec::<SmallVec<[Range; 1]>>::new();
+    let mut prev = None;
+    for (tabstop, range) in tabstops {
+        if prev == Some(tabstop) {
+            let len_1 = ntabstops.len() - 1;
+            ntabstops[len_1].push(range);
+        } else {
+            prev = Some(tabstop);
+            ntabstops.push(smallvec![range]);
+        }
+    }
+
+    if let Some(first) = ntabstops.first() {
+        transaction.with_selection(Selection::new(first.clone(), 0))
```

Comment:
This doesn't work for multiple cursors.

This will only create a single cursor at the tabstop for the primary cursor and delete all other selections.
Instead we should track the offset relative to the start and then compute the new selection (see suggestion above). Then you can compute the selection as follows:

```rust
selection.transform(|range| {
        let cursor = range.cursor(text);
        Range::new(cursor + first.0, cursor + first.1 )
})
```

Reply:
I've added a variant of `Selection::tranform` called `Selection::tranform_iter` that returns a iterator. And with the relative offsets, I think this should handle multiple cursors correctly.